### PR TITLE
fix(java_extractor): preserve dummy output path for modular builds

### LINKED ISSUE #3671: support java modules
#3607 was a first pass at this, but does so in a way that kills caching behavior on compilation units.

### PR DESCRIPTION
**Pull request by shahms**
Fixes #3671 for extraction (at least partially)

### CODE REVIEW

File: kythe/java/com/google/devtools/kythe/extractors/java/JavaCompilationUnitExtractor.java [hunk 1]
```diff
@@ -328,12 +328,27 @@ public CompilationDescription extract(
...
+    // source-compatible with Java 8, check for the support and presence of a
+    // --module-source-path option.
+    EnumSet<Option> remove =
+        hasModuleSourcePath(options)
+            ? EnumSet.of(Option.S, Option.H)
+            : EnumSet.of(Option.D, Option.S, Option.H);
     return ModifiableOptions.of(options)
-        .removeOptions(EnumSet.of(Option.D, Option.S, Option.H))
+        .removeOptions(remove)
+        .replaceOptionValue(Option.D, "/dev/null")
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/extractors/java/JavaCompilationUnitExtractor.java]: Please document what the `/dev/null` does.

Comment (shahms) [reply kythe/java/com/google/devtools/kythe/extractors/java/JavaCompilationUnitExtractor.java]: You mean aside from:
```
    // Option.D needs to remain for module support, but shouldn't point to a concrete path
    // as those are often unique paths which destroy caching.
```

Comment (schroederc) [reply kythe/java/com/google/devtools/kythe/extractors/java/JavaCompilationUnitExtractor.java]: Yes, it's not clear what passing a file to a flag that requires a directory does.

Comment (shahms) [reply kythe/java/com/google/devtools/kythe/extractors/java/JavaCompilationUnitExtractor.java]: Digging in a bit more, I think this needs to be handled in the file manager itself.  "Hilariously" javac check that each od -d, -h, and -s either:

1) aren't specified
2) are specified but don't exist
3) are specified, exist and are a directory

The existence and directory checks are done on the *real* filesystem unconditionally which means they shouldn't be present in the command line.

**Pull request closed without merging**